src/Native/Runtime/threadstore.cpp

```diff
@@ -141,15 +141,21 @@ void ThreadStore::AttachCurrentThread(bool fAcquireThreadStoreLock)
     ASSERT(pAttachingThread->m_ThreadStateFlags == Thread::TSF_Unknown);
 
     ThreadStore* pTS = GetThreadStore();
-    ReaderWriterLock::WriteHolder write(&pTS->m_Lock, fAcquireThreadStoreLock);
-
-    //
-    // Set thread state to be attached
-    //
-    ASSERT(pAttachingThread->m_ThreadStateFlags == Thread::TSF_Unknown);
-    pAttachingThread->m_ThreadStateFlags = Thread::TSF_Attached;
-
-    pTS->m_ThreadList.PushHead(pAttachingThread);
+    {
+        ReaderWriterLock::WriteHolder write(&pTS->m_Lock, fAcquireThreadStoreLock);
+        //
+        // Set thread state to be attached
+        //
+        ASSERT(pAttachingThread->m_ThreadStateFlags == Thread::TSF_Unknown);
+        pAttachingThread->m_ThreadStateFlags = Thread::TSF_Attached;
+
+        pTS->m_ThreadList.PushHead(pAttachingThread);
+    }
+#if defined(CORERT)
+    // REVIEW: should this be done prior to pushing into the thread list.
```

Comment:
It should not be done here at all. You are allocating objects on ConstuctThreadStatics, that can trigger GC - but there is no way for GC to stackwalk from here.